interface/grpc/core/deploy.go

```diff
@@ -29,15 +34,53 @@ func (s *Server) DeployService(stream coreapi.Core_DeployServiceServer) error {
 		err             error
 	)
 
-	sr := newDeployServiceStreamReader(stream)
-	url, err := sr.GetURL()
-	if err != nil {
+	// receive the first message in the stream.
+	if err := sr.RecvMessage(); err != nil {
 		return err
 	}
-	if url != "" {
-		service, validationError, err = s.api.DeployServiceFromURL(url, api.DeployServiceStatusOption(statuses))
+
+	deployOptions := []api.DeployServiceOption{
+		api.DeployServiceStatusOption(statuses),
+	}
+
+	// receive confirmation for a force deploy.
+	// if it's not provided deployment will start anyway.
+	if sr.Confirmation != nil {
+		confirmation := sr.Confirmation.GetValue()
+		deployOptions = append(deployOptions, api.DeployServiceConfirmationOption(func(alias string) bool {
+			return confirmation
+		}))
+	} else {
+		deployOptions = append(deployOptions, api.DeployServiceConfirmationOption(func(alias string) bool {
+			// request for confirmation.
+			if err := stream.Send(&coreapi.DeployServiceReply{
+				Value: &coreapi.DeployServiceReply_RequestConfirmation{RequestConfirmation: alias},
+			}); err != nil {
+				return false
+			}
+
+			// receive the confirmation result.
+			// TODO(ilgooz) add timeout.
+			if err := sr.RecvMessage(); err != nil {
+				return false
+			}
+			if sr.Confirmation == nil {
+				return false
+			}
+			return sr.Confirmation.GetValue()
+		}))
+	}
+
+	if sr.Confirmation != nil {
+		if err := sr.RecvMessage(); err != nil {
+			return err
+		}
+	}
+
+	if sr.URL != "" {
+		service, validationError, err = s.api.DeployServiceFromURL(sr.URL, deployOptions...)
```

Comment:
This functions is hard to read for other developers, as its not clear what's going on here

Reply:
what do you mean?

Reply:
I don't know how exactly explain `it's hard to read`. When you first looking on the function (and you're not an author of it) you don't have an idea what's going on inside without further reading.

Reply:
I don't feel the same. Feel free to put some comments if you wish but I think var/func naming is very good here so it self-explains.